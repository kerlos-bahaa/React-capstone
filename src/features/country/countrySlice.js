import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

import axios from 'axios';

const api = 'https://restcountries.com/v3.1/all';
const apiCurrencyUrl = 'https://countriesnow.space/api/v0.1/countries/currency';

const initialState = {
  country: [],
  countrySecond: [],
  countryDetails: [],
  isLoading: false,
};

export const fetchCoutries = createAsyncThunk('countries/fetchCountries', async () => {
  try {
    const data = await axios.get(api);
    const dataApi = data;
    return dataApi.data;
  } catch (err) {
    return err;
  }
});

export const fetchCoutriesCurrency = createAsyncThunk('countriesCurrency/fetchcountriesCurrency', async () => {
  try {
    const data = await axios.get(apiCurrencyUrl);
    const dataApi = data;
    return dataApi.data;
  } catch (err) {
    return err;
  }
});

const countrySlice = createSlice({
  name: 'country',
  initialState,
  reducers: {
    filterCountry: (state, action) => {
      const filteredState = [...state.countrySecond];
      const receivedCountry = action.payload.toLowerCase();
      const data = filteredState.filter((item) => (
        item.name.common.toLowerCase().includes(receivedCountry)));
      if ((data.length % 2) === 0 || data.length === 1) {
        state.country = [];
        state.country = filteredState;
        state.country = data;
      }
    },
    searchDetails: (state, action) => {
      const filteredSate = [...state.country];
      state.countryDetails = [];
      let val = [];
      const data = filteredSate.filter((item) => item.name.common === action.payload);
      if (data.length > 0) {
        val.push(data);
        state.countryDetails = data;
      } else {
        val = [];
      }
    },
    loading: (state, action) => {
      state.isLoading = action.payload;
    },
  },

  extraReducers(builder) {
    builder.addCase(fetchCoutries.fulfilled, (state, action) => {
      const country = action.payload;
      country.forEach((element) => {
        state.country.push(element);
        state.countrySecond = state.country;
      });
    });
  },
});

export const {
  filterCountry, searchDetails, searchDetailsCurrenc, loading,
} = countrySlice.actions;
export default countrySlice.reducer;
