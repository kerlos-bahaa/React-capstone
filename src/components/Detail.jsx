import { useLocation } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import React, { useEffect } from 'react';
import { searchDetails, loading } from '../features/country/countrySlice';
import SimpleMap from './map';
import IconArrowRightCircle from './IconArrowRightCircle';

const Detail = () => {
  const loc = useLocation();
  const dispatch = useDispatch();
  const { country } = loc.state;
  const data = useSelector((state) => state.country);

  useEffect(() => {
    dispatch(searchDetails(country));
    dispatch(loading(true));
  }, [country, dispatch]);

  return (
    <div>
      {
                data.countryDetails.map((item) => (
                  <div className="details" key={item.name.official}>
                    <div className="dtls">
                      <div className="map">
                        <h3 className="flag-name">
                          {item.name.common.toUpperCase()}
                          <img src={item.flags.png !== undefined ? item.flags.png : 'Flag not avaible'} alt={item.name.official} />
                        </h3>
                        <SimpleMap lat={item.latlng[0]} lng={item.latlng[1]} />
                      </div>
                      <div>
                        <ul className="ul">
                          <li className="crt-details li">COUNTRY DETAILS</li>
                          <li className="liste li">
                            <p>CAPITAL </p>
                            {' '}
                            <p>
                              {item.capital[0] !== undefined ? item.capital[0] : ''}
                              {' '}
                              <IconArrowRightCircle />
                            </p>
                          </li>
                          <li className="pair-red liste li">
                            <p>AREA </p>
                            <p>
                              {' '}
                              {item.area}
                              {' '}
                              km2
                              {' '}
                              <IconArrowRightCircle />
                            </p>
                          </li>
                          <li className="liste li">
                            <p>POPULATION </p>
                            <p>
                              {item.population}
                              {' '}
                              <IconArrowRightCircle />
                            </p>
                          </li>
                          <li className="pair-red liste li">
                            <p>CONTINENT</p>
                            <p>
                              {' '}
                              {item.continents[0] !== undefined ? item.continents[0] : ''}
                              {' '}
                              <IconArrowRightCircle />
                            </p>
                          </li>
                        </ul>
                      </div>
                    </div>
                  </div>
                ))
            }
    </div>
  );
};

export default Detail;
