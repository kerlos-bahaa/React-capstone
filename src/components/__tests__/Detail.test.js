import { MemoryRouter } from 'react-router-dom';
import React from 'react';
import renderer from 'react-test-renderer';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import mockDetails from '../mockDetails';
import Detail from '../Detail';
import { searchDetails } from '../../features/country/countrySlice';

jest.mock('../map');
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => (
    {
      pathname: '/details',
      search: '',
      hash: '',
      state: { country: 'Turks and Caicos Islands' },
      key: '6ma6yc3n',
    }
  ),
}));

const reducer = (state = {
  country: { country: mockDetails, countryDetails: mockDetails },
}) => state;

describe('should test details page', () => {
  it('should test the snapshot of details page', async () => {
    jest.fn(() => searchDetails.mockReturnValue({ countryDetails: mockDetails }));
    const store = configureStore({ reducer });
    const navb = renderer.create(
      <MemoryRouter>
        <Provider store={store}>
          <Detail />
        </Provider>
      </MemoryRouter>,
    ).toJSON();

    expect(navb).toMatchSnapshot();
  });
});
