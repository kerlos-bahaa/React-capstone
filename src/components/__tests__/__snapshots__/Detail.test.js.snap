// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test details page should test the snapshot of details page 1`] = `
<div>
  <div
    className="details"
  >
    <div
      className="dtls"
    >
      <div
        className="map"
      >
        <h3
          className="flag-name"
        >
          JORDAN
          <img
            alt="Hashemite Kingdom of Jordan"
            src="https://flagcdn.com/w320/jo.png"
          />
        </h3>
      </div>
      <div>
        <ul
          className="ul"
        >
          <li
            className="crt-details li"
          >
            COUNTRY DETAILS
          </li>
          <li
            className="liste li"
          >
            <p>
              CAPITAL 
            </p>
             
            <p>
              Amman
               
              <svg
                className="arrow"
                fill="currentColor"
                height="1.2em"
                viewBox="0 0 16 16"
                width="1.2em"
              >
                <path
                  d="M1 8a7 7 0 1014 0A7 7 0 001 8zm15 0A8 8 0 110 8a8 8 0 0116 0zM4.5 7.5a.5.5 0 000 1h5.793l-2.147 2.146a.5.5 0 00.708.708l3-3a.5.5 0 000-.708l-3-3a.5.5 0 10-.708.708L10.293 7.5H4.5z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </li>
          <li
            className="pair-red liste li"
          >
            <p>
              AREA 
            </p>
            <p>
               
              89342
               
              km2
               
              <svg
                className="arrow"
                fill="currentColor"
                height="1.2em"
                viewBox="0 0 16 16"
                width="1.2em"
              >
                <path
                  d="M1 8a7 7 0 1014 0A7 7 0 001 8zm15 0A8 8 0 110 8a8 8 0 0116 0zM4.5 7.5a.5.5 0 000 1h5.793l-2.147 2.146a.5.5 0 00.708.708l3-3a.5.5 0 000-.708l-3-3a.5.5 0 10-.708.708L10.293 7.5H4.5z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </li>
          <li
            className="liste li"
          >
            <p>
              POPULATION 
            </p>
            <p>
              10203140
               
              <svg
                className="arrow"
                fill="currentColor"
                height="1.2em"
                viewBox="0 0 16 16"
                width="1.2em"
              >
                <path
                  d="M1 8a7 7 0 1014 0A7 7 0 001 8zm15 0A8 8 0 110 8a8 8 0 0116 0zM4.5 7.5a.5.5 0 000 1h5.793l-2.147 2.146a.5.5 0 00.708.708l3-3a.5.5 0 000-.708l-3-3a.5.5 0 10-.708.708L10.293 7.5H4.5z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </li>
          <li
            className="pair-red liste li"
          >
            <p>
              CONTINENT
            </p>
            <p>
               
              Asia
               
              <svg
                className="arrow"
                fill="currentColor"
                height="1.2em"
                viewBox="0 0 16 16"
                width="1.2em"
              >
                <path
                  d="M1 8a7 7 0 1014 0A7 7 0 001 8zm15 0A8 8 0 110 8a8 8 0 0116 0zM4.5 7.5a.5.5 0 000 1h5.793l-2.147 2.146a.5.5 0 00.708.708l3-3a.5.5 0 000-.708l-3-3a.5.5 0 10-.708.708L10.293 7.5H4.5z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
